File: sdk/csharp/libraries/microsoft.bot.builder.solutions.skills/SkillMiddleware.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Skills.Models;
using Microsoft.Bot.Schema;

namespace Microsoft.Bot.Builder.Skills
{
    /// <summary>
    /// The Skill middleware is responsible for processing Skill mode specifics, for example the skillBegin event used to signal the start of a skill conversation.
    /// </summary>
    public class SkillMiddleware : IMiddleware
    {
        private UserState _userState;
        private ConversationState _conversationState;
        private IStatePropertyAccessor<DialogState> _dialogState;

        public SkillMiddleware(UserState userState, ConversationState conversationState, IStatePropertyAccessor<DialogState> dialogState)
        {
            _userState = userState;
            _conversationState = conversationState;
            _dialogState = dialogState;
        }

        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
        {
            var activity = turnContext.Activity;
            if (activity != null && activity.Type == ActivityTypes.Event)
            {
                if (activity.Name == SkillEvents.CancelAllSkillDialogsEventName)
                {
                    await _dialogState.DeleteAsync(turnContext).ConfigureAwait(false);
                    await _conversationState.ClearStateAsync(turnContext).ConfigureAwait(false);
                    await _conversationState.SaveChangesAsync(turnContext, force: true).ConfigureAwait(false);
                    return;
                }
            }

            await next(cancellationToken).ConfigureAwait(false);
        }
    }
}

File: sdk/typescript/libraries/botbuilder-solutions/src/skills/skillMiddleware.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { ConversationState, Middleware, StatePropertyAccessor, TurnContext, UserState } from 'botbuilder';
import { DialogState } from 'botbuilder-dialogs';
import { Activity, ActivityTypes } from 'botframework-schema';

/**
 * The Skill middleware is responsible for processing Skill mode specifics,
 * for example the skillBegin event used to signal the start of a skill conversation.
 */
export class SkillMiddleware implements Middleware {
    private readonly userState: UserState;
    private readonly conversationState: ConversationState;
    private readonly dialogState: StatePropertyAccessor<DialogState>;

    public constructor(
        userState: UserState,
        conversationState: ConversationState,
        dialogState: StatePropertyAccessor<DialogState>
    ) {
        this.userState = userState;
        this.conversationState = conversationState;
        this.dialogState = dialogState;
    }

    public async onTurn(turnContext: TurnContext, next: () => Promise<void>): Promise<void> {
        const activity: Activity = turnContext.activity;
        if (activity !== undefined && activity.type === ActivityTypes.EndOfConversation) {
            await this.dialogState.delete(turnContext);
            await this.conversationState.clear(turnContext);
            await this.conversationState.saveChanges(turnContext, true);

            return;
        }

        await next();
    }
}
